js/src/gtag-events/utils.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * Internal dependencies
+ */
+import { sendToGroup } from './constants';
+
+/* global glaGtagData */
+
+/**
+ * Track an event using the global gtag function.
+ *
+ * @param {string} eventName
+ * @param {Object} eventParams
+ */
+export const trackEvent = ( eventName, eventParams ) => {
+	if ( typeof gtag !== 'function' ) {
+		throw new Error( 'Function gtag not implemented.' );
+	}
+	// eslint-disable-next-line no-console
+	console.log( `Tracking event ${ eventName }` );
+	window.gtag( 'event', eventName, {
+		send_to: sendToGroup,
+		...eventParams,
+	} );
+};
+
+/**
+ * Track an add_to_cart event.
+ *
+ * @param {Object} product
+ * @param {number} quantity
+ */
+export const trackAddToCartEvent = ( product, quantity = 1 ) => {
+	trackEvent( 'add_to_cart', {
+		ecomm_pagetype: 'cart',
+		event_category: 'ecommerce',
+		items: [ getItemObject( product, quantity ) ],
+	} );
+};
+
+/**
+ * Formats data into a cart Item object.
+ *
+ * @param {Object} product
+ * @param {number} quantity
+ * @return {Object} Item object.
+ */
+export const getItemObject = ( product, quantity ) => {
+	const item = {
+		id: 'gla_' + product.id,
+		quantity,
+		google_business_vertical: 'retail',
+	};
+
+	if ( product.name ) {
+		item.name = product.name;
+	}
+
+	if ( 'categories' in product && product.categories.length ) {
```

Comment:
💅  

`if (product?.categories?.length ) {`

also works

❓  Wondering if it's possible to have category without name. Since in that case `item.category` would be undefined.

Reply:
The categories are populated when the hook `experimental__woocommerce_blocks-cart-add-item` gets triggered. Each category there will have a name. For older add_to_cart buttons we don't track the item categories (since it's optional and it requires traversing and storing additional data per product).

Reply:
This is also code copied from WooCommerce blocks. I understand that we have quite high requirements on the backend for JS / browsers. Do we have the same on the frontend, should I be worried at all that optional chaining might not work on some browsers? Have the same question about the exponentiation operator.

Reply:
> Do we have the same on the frontend, should I be worried at all that optional chaining might not work on some browsers? 

We  (wp-scripts) use [babel](https://babeljs.io/) behind the scenes when we build (compile) the code. In brief this converts code using new syntax to be compatible. Besides that, in this case .? chaining operator is [fully supported since 2020 in most browsers.  ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining)

PS. If you're curious you can check `js/build/gtag-events.js` to see how the code compiles.

Reply:
>  Have the same question about the exponentiation operator.

Same answer. Actually it has[ full support since 2016](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Exponentiation) (However, I did't use that one before)